src/main/java/io/github/opencubicchunks/cubicchunks/core/server/VanillaNetworkHandler.java

```diff
@@ -125,38 +139,58 @@ public VanillaNetworkHandler(WorldServer world) {
+        return playerOffsetsC2S.getOrDefault(player, CubePos.ZERO);
     }
 
     public void sendCubeLoadPackets(Collection<? extends ICube> cubes, EntityPlayerMP player) {
+        if (!CubicChunksConfig.allowVanillaClients) {
+            return;
+        }
+        sendCubeLoadPackets(cubes, player, getPlayerOffsetS2C(player));
+    }
+
+    public void sendCubeLoadPackets(Collection<? extends ICube> cubes, EntityPlayerMP player, CubePos offset) {
         if (!CubicChunksConfig.allowVanillaClients) {
             return;
         }
         Map<ChunkPos, List<ICube>> columns = cubes.stream().collect(Collectors.groupingBy(c -> c.getCoords().chunkPos()));
         for (Map.Entry<ChunkPos, List<ICube>> chunkPosListEntry : columns.entrySet()) {
             ChunkPos pos = chunkPosListEntry.getKey();
             List<ICube> column = chunkPosListEntry.getValue();
-            SPacketChunkData chunkData = constructChunkData(pos, column, getPlayerOffset(player), world.provider.hasSkyLight());
+            clientLoadedChunkPositions.computeIfAbsent(player, p -> new Object2ObjectOpenHashMap<>())
+                    .put(new ChunkPos(pos.x + offset.getX(), pos.z + offset.getZ()), pos);
+            SPacketChunkData chunkData = constructChunkData(pos, column, offset, world.provider.hasSkyLight());
             player.connection.sendPacket(chunkData);
         }
     }
 
-    public void sendFullCubeLoadPackets(Collection<? extends ICube> cubes, EntityPlayerMP player) {
+    public void sendFullCubeLoadPackets(Collection<? extends ICube> cubes, EntityPlayerMP player, Map<ChunkPos, ChunkPos> loadedChunks, CubePos oldOffset, CubePos newOffset) {
         if (!CubicChunksConfig.allowVanillaClients) {
             return;
         }
         Map<Chunk, List<ICube>> columns = cubes.stream().collect(Collectors.groupingBy(ICube::getColumn));
         for (Map.Entry<Chunk, List<ICube>> chunkPosListEntry : columns.entrySet()) {
             Chunk chunk = chunkPosListEntry.getKey();
             List<ICube> column = chunkPosListEntry.getValue();
-            SPacketChunkData chunkData = constructFullChunkData(chunk, column, getPlayerOffset(player), world.provider.hasSkyLight());
-            SPacketUnloadChunk unload = new SPacketUnloadChunk(chunk.x, chunk.z);
-            player.connection.sendPacket(unload);
+            SPacketChunkData chunkData = constructFullChunkData(chunk, column, newOffset, world.provider.hasSkyLight());
+
+            ChunkPos clientChunkPosNew = new ChunkPos(chunk.x + newOffset.getX(), chunk.z + newOffset.getZ());
+            ChunkPos clientChunkPos = new ChunkPos(chunk.x + oldOffset.getX(), chunk.z + oldOffset.getZ());
+            ChunkPos serverChunkPos = loadedChunks.get(clientChunkPos);
+            if ((serverChunkPos != null && serverChunkPos.x == chunk.x && serverChunkPos.z == chunk.z)
+                || loadedChunks.containsKey(clientChunkPos = clientChunkPosNew)) {
```

Comment:
How can this code be reached without the loadedChunks check being true? The cubes in the map should be only the cubes that have already been sent to clients.

Honestly this seems like a very hacky and broken workaround for a deeper issue of handling overlapping but not matching areas, where you may actually end up sending a chunk packet for a position that is not yet unloaded. A better solution would be to send all the unload packets first, and then all the load packets.

Reply:
I have to agree, this is definitely a hacky workaround and honestly I'm not sure why I didn't think of just sending all the chunk unloads at once. I'll rewrite it to do that.